Topology: Add high-pass EQ with gain into DMIC capture pipeline

singalsu opened the pull request:
This patch set for topology and EQ tool contains

- Updates to topology to increase default gain of DMIC capture by +20 dB
- Decrease volume max. gain by 10 dB 30 -> 20 dB, lower number of steps 80 -> 70. Previous pipeline max. possible gain was +30 dB. Now it is +40 dB. I did the 10 dB drop because I felt +50 dB would have been excessive.
- Prevent capture of 2ch from 4ch configured pipeline. Attempt to capture 2ch created corrupted audio without any errors returned to arecord.
- Update paths into script that generates example IIR EQs and add 16 kHz rate high-pass

See commit messages for detailed description

plbossart commented:
@singalsu thanks, this is a nice improvement. see below the results on my WHL device. seems to work much better than anything I tried before :-)

![pr1275-2](https://user-images.githubusercontent.com/1149035/62966481-c3f37080-bdcc-11e9-8628-b22dcf9d3a17.png)
[plb.wav.gz](https://github.com/thesofproject/sof/files/3498313/plb.wav.gz)

RanderWang commented:
@ClarexZhou  Can you check the effect ?

ClarexZhou commented:
Test on CML HDA notebook. DMIC6 and DMIC7 default volume is 100.  PGA10 and PGA11 volume max is changed to 70, previous is 80.   Captured audio is loud than before in same volume level.  

 bug 1131 is fixed, audio captured in 50% and output can be heard.

singalsu commented:
@RanderWang Do you have ideas why the ICL RVP CI fails?

singalsu commented:
@plbossart Nice, thanks for testing and confirming it's OK with four microphones. I have here a two microphones device.

singalsu commented:
I updated this PR. There's no changes into topology files but I added the EQ instance identifiers placement into the IIR example generator script so the used EQ configs can be used without manual edits into m4 data files. I forgot it from the previous version of PR. Also I added commit description text into first commit.

singalsu commented:
SOFCI TEST

cujomalainey commented on tools/topology/sof-hda-generic.m4:
```diff
@@ -24,6 +24,8 @@ ifelse(CHANNELS, `0', ,
 define(DMIC_PIPELINE_48k_ID, `10')
 define(DMIC_PIPELINE_16k_ID, `11')
 
+DEBUG_START
```
is this supposed to be here?

singalsu replied:
I'm not sure if there's any overhead of having this. I'll check if it impact *.tplg files. With DEBUG_START - DEBUG_END we are able to plot the topology as described in

https://thesofproject.github.io/latest/developer_guides/topology/topology.html#debugging-topology

Some topologies have them and some do not.

singalsu replied:
But this fix/enhance does not depend on the DEBUGs so I think I should remove them from this PR.

juimonen replied:
the DEBUG should not propagate to .conf or .tplg file if you are just using the m4 and alsa tools as defined in makefiles. As you said it is just a helper to plot the graph. And you can just remove it as it is not part of this PR. Anyway, good to spread the info how to plot the topology :)

singalsu replied:
Done!

cujomalainey commented on tools/tune/eq/example_iir_eq.m:
```diff
@@ -192,6 +194,33 @@ function example_iir_eq()
+comment = '50 Hz second order high-pass and +20 dB gain';
+
+%% Design IIR high-pass
+eq_hp = hp_iir_eq(16e3);
```
is this is just using a biquad to do the dcblocking effect? Just wondering about porting over our first order solution as well so we can port our tuning configs straight over when the time comes.

singalsu replied:
Yes it's the simple N biquads in series filter. It's now optimized with intrinsics for HiFi3 so there's not much overhead. This is just a single biquad. A 1st order version would work here as well. I've configured this to provide also the 20 dB gain since today we cannot define in topology the initial volume gain to be other than firmware default 0 dB.

singalsu replied:
If we will able set the initial gain for volume instances in topology I'd revert the gain of high-pass to 0 dB. With this topology the user is not able lower the gain if the capture would clip under high sound pressure levels. But since the devices that use this topology are notebooks I'd assume it's not an issue in practice.

The initial volume passing to firmware from topology would need an ABI bump. It's one idea to address the issue with default gains.

Also the volume scale could be defined to be e.g. +20 .. +50 dB then volume comp would use the lowest gain in scale as default. But then gain could not be adjusted to any lower than +20 dB. So This is what I proposed then as full rationale explained!

cujomalainey commented on tools/topology/sof/pipe-eq-capture-16khz.m4:
```diff
@@ -0,0 +1,108 @@
+# Capture EQ Pipeline and PCM, 16 kHz
```
is the reason this is fixed to 16kHz simply because that is what the tuning parameters were set for?

singalsu replied:
I think it's Intel FW legacy to setup the DAI for 48 kHz and 16 kHz configurations and we are following it. The 16 kHz is a typical sample rate for automatic speech recognition. Though in SOF case it could be a different rate to match needs.

singalsu commented:
This PR needed a rebase due to conflicts. I've tested this to work again OK in WHL laptop.

singalsu commented:
Once required m4 file was missing from commit - fixed now.
